gh-91053: Add an optional callback that is invoked whenever a function is modified

mpage opened the pull request:
JIT compilers may need to invalidate compiled code when a function is modified (e.g. if its code object is modified). This adds the ability to set a callback that, when set, is called each time a function is modified.

<!-- gh-issue-number: gh-91053 -->
* Issue: gh-91053
<!-- /gh-issue-number -->

mpage committed:
Add an optional callback that is invoked whenever a function is modified

JIT compilers may need to invalidate compiled code when a function is
modified (e.g. if its code object is modified). This adds the ability
to set a callback that, when set, is called each time a function is
modified.

erlend-aasland commented on Include/cpython/funcobject.h:
```diff
@@ -131,6 +131,52 @@ PyAPI_DATA(PyTypeObject) PyStaticMethod_Type;
 PyAPI_FUNC(PyObject *) PyStaticMethod_New(PyObject *);
 
+#define FOREACH_FUNC_EVENT(V) \
+    V(CREATED)                \
```
The other added constants use the imperative mood. Suggesting to be consistent and also use the imperative mood here.

erlend-aasland commented on Include/cpython/funcobject.h:
```diff
@@ -131,6 +131,52 @@ PyAPI_DATA(PyTypeObject) PyStaticMethod_Type;
+    V(MODIFY_KWDEFAULTS)
+
+typedef enum {
+    #define DEF_EVENT(EVENT) PyFunction_EVENT_##EVENT,
+    FOREACH_FUNC_EVENT(DEF_EVENT)
+    #undef DEF_EVENT
```
IMO, it unneccesary messy to use two macros to create a list of five enums; I'd simply spell them out. I guess it is fine to keep it this way also, but at least `#undef FOREACH_FUNC_EVENT` when you're done with it.

mpage replied:
This is intentionally left defined to provide others a means to programmatically apply a macro to each function event at compile time (https://en.wikipedia.org/wiki/X_Macro). This is used in `func_events.c` to expose string constants for each function event.

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+static PyObject *pyfunc_watchers[NUM_WATCHERS];
+static int watcher_ids[NUM_WATCHERS] = {-1, -1};
+
+static PyObject *get_id(PyObject *obj) {
```
PEP7-nit:

```suggestion
static PyObject *
get_id(PyObject *obj)
{
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+static int watcher_ids[NUM_WATCHERS] = {-1, -1};
+
+static PyObject *get_id(PyObject *obj) {
+    PyObject *builtins = PyEval_GetBuiltins();
```
Please add a comment that this is a borrowed reference.

```suggestion
    PyObject *builtins = PyEval_GetBuiltins();  // borrowed ref.
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+static int
+call_pyfunc_watcher(PyObject *watcher, PyFunction_WatchEvent event, PyFunctionObject *func, PyObject *new_value)
```
PEP 7-nit:

```suggestion
call_pyfunc_watcher(PyObject *watcher, PyFunction_WatchEvent event,
                    PyFunctionObject *func, PyObject *new_value)
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+            Py_DECREF(new_value);
+            return -1;
+        }
+    } else {
```
PEP7

```suggestion
    }
    else {
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+static int
+first_watcher_callback(PyFunction_WatchEvent event, PyFunctionObject *func, PyObject *new_value)
```
PEP7: wrap at 79 chars

```suggestion
first_watcher_callback(PyFunction_WatchEvent event, PyFunctionObject *func,
                       PyObject *new_value)
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+static int
+second_watcher_callback(PyFunction_WatchEvent event, PyFunctionObject *func, PyObject *new_value)
```
PEP7: wrap at 79 chars

```suggestion
second_watcher_callback(PyFunction_WatchEvent event, PyFunctionObject *func,
                        PyObject *new_value)
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+
+static PyFunction_WatchCallback watcher_callbacks[NUM_WATCHERS] = {
+  first_watcher_callback,
+  second_watcher_callback
```
PEP7: 4 space indent

```suggestion
    first_watcher_callback,
    second_watcher_callback,
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+        return NULL;
+    }
+    pyfunc_watchers[idx] = func;
+    Py_INCREF(func);
```
Shortcut:

```suggestion
    pyfunc_watchers[idx] = Py_NewRef(func);
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+      if (watcher_ids[i] == wid) {
+          idx = i;
+          break;
+      }
```
PEP7: 4 spaces indent

```suggestion
        if (watcher_ids[i] == wid) {
            idx = i;
            break;
        }
```

erlend-aasland commented on Objects/funcobject.c:
```diff
@@ -3,12 +3,57 @@
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    if (watcher_id < 0 || watcher_id >= FUNC_MAX_WATCHERS) {
+        PyErr_Format(PyExc_ValueError, "Invalid func watcher ID %d", watcher_id);
```
Convention (at least in this file) is to start the exception message with a lower-case letter. Also, wrap at 79 chars.

```suggestion
        PyErr_Format(PyExc_ValueError, "invalid func watcher ID %d",
                     watcher_id);
```

erlend-aasland commented on Objects/funcobject.c:
```diff
@@ -3,12 +3,57 @@
+    }
+    PyErr_SetString(PyExc_RuntimeError, "no more func watcher IDs available");
+    return -1;
+
```
```suggestion
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+clear_watcher(PyObject *self, PyObject *watcher_id_obj)
+{
+    long watcher_id = PyLong_AsLong(watcher_id_obj);
+    if ((watcher_id < INT_MIN) || (watcher_id > INT_MAX)) {
```
Watcher id's less than 0 are not allowed (unless I misread something), so the lower bounds check should be `watcher_id < 0`, right?

erlend-aasland replied:
Also, the upper bound should be `FUNC_MAX_WATCHERS` inclusive.

mpage replied:
This is to catch numeric errors when casting from types of potentially different sizes. The bounds checks on watcher ids happen in `PyFunction_ClearWatcher`.

erlend-aasland commented on Objects/funcobject.c:
```diff
@@ -3,12 +3,57 @@
+        return -1;
+    }
+    if (!interp->func_watchers[watcher_id]) {
+        PyErr_Format(PyExc_ValueError, "No func watcher set for ID %d", watcher_id);
```
Ditto

```suggestion
        PyErr_Format(PyExc_ValueError, "no func watcher set for ID %d",
                     watcher_id);
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+};
+
+int
+_PyTestCapi_Init_FuncEvents(PyObject *m) {
```
PEP 7

```suggestion
_PyTestCapi_Init_FuncEvents(PyObject *m)
{
```

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -0,0 +1,168 @@
+
+    /* Expose each event as an attribute on the module */
+#define ADD_EVENT(event)  \
+    if (add_event(m, "PYFUNC_EVENT_" #event, PyFunction_EVENT_##event)) { \
```
FYI, you can get rid of the whole `add_event` helper if you use `PyModule_AddIntMacro`.

mpage replied:
I ended up keeping things as-is so that I can expose the constants to Python as `SCREAMING_SNAKE_CASE` rather than a mix of that and `PascalCase`.

erlend-aasland commented on Doc/c-api/function.rst:
```diff
@@ -118,3 +118,56 @@ There are a few functions specific to Python functions.
+   Register *callback* as a function watcher for the current interpreter. Returns
+   an id which may be passed to :c:func:`PyFunction_ClearWatcher`. In case
+   of error (e.g. no more watcher IDs available), return ``-1`` and set an
+   exception.
```
Nit: semantic line-breaks will make this paragraph more neatly formatted. Personal preference only; you may ignore this suggestion. Also, consistently use imperative mood.

```suggestion
   Register *callback* as a function watcher for the current interpreter.
   Return an id which may be passed to :c:func:`PyFunction_ClearWatcher`.
   In case of error (e.g. no more watcher IDs available),
   return ``-1`` and set an exception.
```

erlend-aasland replied:
Also, throughout the PR you use both lower-case `id` and upper-case `ID` (docs, comments, exception messages, etc.). I'd consider sticking to one of them for consistency.

erlend-aasland commented on Doc/c-api/function.rst:
```diff
@@ -118,3 +118,56 @@ There are a few functions specific to Python functions.
+   Clear watcher identified by *watcher_id* previously returned from
+   :c:func:`PyFunction_AddWatcher` for the current interpreter. Return ``0`` on
+   success or ``-1`` on error (e.g.  if the given *watcher_id* was never
+   registered.)
```
I'd append _"and set an exception"_ for the error description. For example:

```suggestion
   :c:func:`PyFunction_AddWatcher` for the current interpreter.
   Return ``0`` on success, or ``-1`` and set an exception on error
   (e.g.  if the given *watcher_id* was never registered.)
```

erlend-aasland commented on Doc/c-api/function.rst:
```diff
@@ -118,3 +118,56 @@ There are a few functions specific to Python functions.
+
+   Enumeration of possible function watcher events: ``PyFunction_EVENT_CREATED``,
+   ``PyFunction_EVENT_DESTROY``, ``PyFunction_EVENT_MODIFY_CODE``,
+   ``PyFunction_EVENT_MODIFY_DEFAULTS``, or ``PyFunction_EVENT_MODIFY_KWDEFAULTS``.
```
Consider formatting these as a list.

erlend-aasland commented on Doc/c-api/function.rst:
```diff
@@ -118,3 +118,56 @@ There are a few functions specific to Python functions.
+
+   If *event* is ``PyFunction_EVENT_CREATED`` or ``PyFunction_EVENT_DESTROY``
+   then *new_value* will be ``NULL``. Otherwise, *new_value* will hold a
+   borrowed reference to the new value that is about to be stored in *func* for
```
Consider adding a reference to ```:term:`borrowed reference` ```.

erlend-aasland commented on Objects/funcobject.c:
```diff
@@ -3,12 +3,57 @@
+            return i;
+        }
+    }
+    PyErr_SetString(PyExc_RuntimeError, "no more func watcher IDs available");
+    return -1;
```
This branch is not tested.

erlend-aasland reviewed (commented):
Looks good. I added a bunch of PEP 7 comments because we should (almost) always strive to follow the style guide, both for new code and (in most cases) when modifying existing code. I also added some comments regarding docs, inconsistent naming of the constants, and some C API suggestions.

Questions:
- `PyFunction_SetDefaults` should emit a `PyFunction_EVENT_MODIFY_DEFAULTS` event, no?
- `PyFunction_SetKwDefaults` should emit a `PyFunction_EVENT_MODIFY_KWDEFAULTS` event, no?
- Why no events for `func_closure` and `func_annotations` modifications?

Also, please add the Python test code to `Lib/test/test_capi.py` instead of adding a new `Lib/test/test_func_events.py`, like the dict watchers did.

mpage committed:
Use imperative mood for all enum values

mpage committed:
Update docs

- Imperative mood
- Semantic line breaks

mpage committed:
Conform to PEP-7

mpage committed:
Dispatch events when {kw}defaults are modified via the C-API

mpage committed:
Move tests into test_capi

mpage committed:
Add a test for allocating too many watchers

mpage committed:
Trigger events when {kw}defaults are changed via the C-API

erlend-aasland commented on Modules/_testcapi/func_events.c:
```diff
@@ -1,11 +1,16 @@
-static PyObject *get_id(PyObject *obj) {
-    PyObject *builtins = PyEval_GetBuiltins();
+static
+PyObject *get_id(PyObject *obj)
```
PEP 7 nit

```suggestion
static PyObject *
get_id(PyObject *obj)
```

mpage replied:
Do you use an automatic formatter like `clang-format`? If so, would you mind sharing your config?

erlend-aasland replied:
Sorry, I don't use an automatic formatter for CPython; I've got PEP 7 in my fingers :) I know there's been discussions about having a PEP 7 check in the CI or doing some kind of automatic formatting, but those ideas have been rejected every time.

See [this Discourse topic](https://discuss.python.org/t/pep-proposal-automatically-formatting-the-cpython-code/5603) for one of the most recent discussions.

erlend-aasland replied:
Paul Ganssle has a `clang-format` config that approximate PEP 7: https://gist.github.com/pganssle/0e3a5f828b4d07d79447f6ced8e7e4db

erlend-aasland reviewed (commented):
Thanks for making the changes! I've got one last minor style nit. Also please merge in main and resolve the conflicts.

mpage committed:
More PEP-7

mpage committed:
Merge branch 'main' into func-watchers

mpage committed:
Merge branch 'main' into func-watchers

markshannon commented on Objects/funcobject.c:
```diff
@@ -3,12 +3,58 @@
+handle_func_event(PyFunction_WatchEvent event, PyFunctionObject *func, PyObject *new_value)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    for (int i = 0; i < FUNC_MAX_WATCHERS; i++) {
```
If we add a bit vector we can skip the iteration when there are no watchers.
Maybe put that test in an inline function and call this function only if there are watchers?

carljm commented on Doc/c-api/function.rst:
```diff
@@ -118,3 +118,59 @@ There are a few functions specific to Python functions.
+   after `func` has been fully initialized. Otherwise, the callback is invoked
+   before the modification to *func* takes place, so the prior state of *func*
+   can be inspected.
+
```
I guess we should add a note here that optimized execution of Python code is permitted to eliminate creation of a function object if possible, and this is not considered a semantic change, even though it is observable if a function watcher is registered.

erlend-aasland commented on Modules/_testcapi/parts.h:
```diff
@@ -27,6 +27,7 @@
 int _PyTestCapi_Init_Vectorcall(PyObject *module);
 int _PyTestCapi_Init_Heaptype(PyObject *module);
 int _PyTestCapi_Init_Unicode(PyObject *module);
+int _PyTestCapi_Init_FuncEvents(PyObject *module);
```
FYI, I recently split out the dict and type watcher tests into `Modules/_testcapi/watchers.c` and `Lib/test/test_capi/test_watchers.py`. Would you mind redirecting your test code to those files? Sorry for the inconvenience.

mpage committed:
Merge branch 'main' into func-watchers

mpage committed:
Document that the runtime is free to optimize away function creation

mpage committed:
Add a bit vector to optimize watcher dispatch

A bit is set in the bit vector iff there is a watcher set at the
corresponding offset in the watcher array. Only notify watchers
if at least one bit is set.

erlend-aasland committed:
Sync with main

erlend-aasland committed:
Revert added newlines to VS project files
